datafusion/common/src/tree_node.rs

```diff
@@ -555,18 +594,15 @@ impl<T: DynTreeNode + ?Sized> TreeNode for Arc<T> {
     {
         let children = self.arc_children();
         if !children.is_empty() {
-            let t = children.into_iter().map_till_continue_and_collect(f)?;
-            // TODO: Currently `assert_eq!(t.transformed, t2.transformed)` fails as
-            //  `t.transformed` quality comes from if the transformation closures fill the
-            //   field correctly.
-            //  Once we trust `t.transformed` we can remove the additional check in
-            //  `with_new_arc_children()`.
-            let arc_self = Arc::clone(&self);
-            let t2 = self.with_new_arc_children(arc_self, t.data)?;
-
-            // Propagate up `t.transformed` and `t.tnr` along with the node containing
-            // transformed children.
-            Ok(Transformed::new(t2.data, t.transformed, t.tnr))
+            let new_children = children.into_iter().map_until_stop_and_collect(f)?;
+            // Propagate up `new_children.transformed` and `new_children.tnr`
+            // along with the node containing transformed children.
+            if new_children.transformed {
```

Comment:
Isn't this very risky? You now are relying on if the transformation closures filled the `transformed` flag correctly. What if they don't? So far noone used that flag for anything and noone checked if the flag is correcty filled. If a closure changes a child node but reports transformed=false (incorrectly) then the node's child will not be changed.
And such incorrect cloures can exists not only in Apache DataFusion code, but anything that depends on DataFusion...

I think it would be safer to always call `let t2 = self.with_new_arc_children(arc_self, t.data)?;`. What I'm undecided about is if we should return:
- `Ok(Transformed::new(t2.data, t.transformed, t.tnr))`
- or `Ok(Transformed::new(t2.data, t2.transformed, t.tnr))`
- or `Ok(Transformed::new(t2.data, t.transformed || t2.transformed, t.tnr))`

Reply:
I can understand your concern, but we are providing an API and I think we should assume it is used correctly. If we provide a good documentation, I don't think there will be any problems. It is clear and not hard to understand, so I don't think someone would return modified data as transformed::no. Carrying this information along would be a burden if we're not going to use it.

Maybe we can change the default value of the methods that create new objects to transformed::yes, or we can close the direct access to data, and force users to modify it through a method (`set_data()`), and when this method is run at least once, we can automatically set the transformed to yes.

Reply:
All right, let's run the tests and see if the current closures use the flag correctly.

BTW, why is `ConcreteTreeNode` different? As far as I see you didn't apply the `if new_children.transformed` check there..

Reply:
`tnr` needs to come from `new_children`, returning it from `self.with_new_arc_children(arc_self, new_children.data)` will not work.

How about:
```
            if new_children.transformed {
                new_children.map_data(|data| {
                    let arc_self = Arc::clone(&self);
                    self.with_new_arc_children(arc_self, data).map(|t| t.data)
                })
            } else {
                Ok(Transformed::no(self))
            }
```

Reply:
> BTW, why is `ConcreteTreeNode` different? As far as I see you didn't apply the `if new_children.transformed` check there..

You are right. I thought `with_new_children` already does that propagation from children, but it only checks plans, not the data. I am updating it.

Reply:
> > `tnr` needs to come from `new_children`, returning it from `self.with_new_arc_children(arc_self, new_children.data)` will not work.
> 
> if `new_children.transformed` is true, I believe there is no possibility of that `self.with_new_arc_children(arc_self, new_children.data)` returns transformed::no. Therefore I thought the existing implementation is safe.

I'm not talking about `.transformed` but about the `.tnr` field of the result. `self.with_new_arc_children(arc_self, new_children.data)` always returns `TreeNodeRecursion::Continue` so we need to return `.trn` from `new_children`.